libhttpserver/server.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libhttpserver
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/golang-lru"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/libfs"
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/keybase/kbfs/libmime"
+	"github.com/keybase/kbfs/tlf"
+)
+
+const tokenCacheSize = 64
+const fsCacheSize = 64
+
+// Server is a local HTTP server for serving KBFS content to the front end.
+type Server struct {
+	config libkbfs.Config
+	server *libkb.HTTPSrv
+
+	tokens *lru.Cache
+	fs     *lru.Cache
+}
+
+var _ libkbfs.LocalHTTPServer = (*Server)(nil)
+
+const tokenByteSize = 16
+
+// NewToken implements the libkbfs.LocalHTTPServer interface..
+func (s *Server) NewToken() (token string, err error) {
+	buf := make([]byte, tokenByteSize)
+	if _, err = rand.Read(buf); err != nil {
+		return "", err
+	}
+	token = hex.EncodeToString(buf)
+	s.tokens.Add(token, nil)
+	return token, nil
+}
+
+func (s *Server) handleInvalidToken(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusForbidden)
+	io.WriteString(w, `
+    <html>
+        <head>
+            <title>KBFS HTTP Token Invalid</title>
+        </head>
+        <body>
+            token invalid
+        </body>
+    </html>
+    `)
+}
+
+func (s *Server) handleBadRequest(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func (s *Server) getHTTPFileSystem(ctx context.Context, requestPath string) (
+	toStrip string, fs http.FileSystem, err error) {
+	fields := strings.Split(requestPath, "/")
+	if len(fields) < 3 {
+		return "", nil, errors.New("bad path")
+	}
+
+	tlfType, err := tlf.ParseTlfType(fields[0])
+	if err != nil {
+		return "", nil, err
+	}
+
+	tlfHandle, err := libkbfs.GetHandleFromFolderNameAndType(ctx,
+		s.config.KBPKI(), s.config.MDOps(), fields[1], tlfType)
+	if err != nil {
+		return "", nil, err
+	}
+
+	toStrip = path.Join(fields[0], fields[1])
+
+	fav := tlfHandle.ToFavorite()
+	if fsCached, ok := s.fs.Get(fav); ok {
```

Comment:
It might be better to do the caching for `toStrip` , rather than the handle, because parsing and resolving the handle is kind of expensive.  Since all the calls will be coming in from the front end, where the names are already normalized, there shouldn't be any downside to it.